### PR TITLE
Pass SessionOptions to XnnpackProviderFactoryCreator.

### PR DESCRIPTION
**Pull request by wejoncy**
### Description
To pass session_options to Xnnpack EP via `XnnpackProviderFactoryCreator` for Initializing xnnpack's threadpool.

If you want to use different threadpool size or even disable xnnpack's threadpool, just setting intra_threadpool to 1 by xnnpack EP's provider_options.

### CODE REVIEW

File: include/onnxruntime/core/framework/execution_provider.h [hunk 1]
```diff
@@ -299,6 +300,14 @@ class IExecutionProvider {
   /** Does the EP support concurrent calls to InferenceSession::Run to execute the model.
    */
   virtual bool ConcurrentRunSupported() const { return true; }
+  /** Some session option values (default or user provided) may not work with some EPs.
+   * The EP should alter the SessionOptions object fit its implementation to avoid potential crash.
```
Comment (skottmckay) [inline include/onnxruntime/core/framework/execution_provider.h:304]: This probably needs a design discussion. Not a huge fan of letting all EPs alter the session options in arbitrary ways. What if two conflict in their requirements? Whichever runs second is going to win, and one will always be broken.

I can understand EPs being able to validate whether they can run with the session options and report an issue. I also prefer returning an error instead of silently 'fixing' an invalid option as that way the user knows we're not running in the way they might think we are. 

This method also seems to be serving two purposes. One is to allow the EP to validate the options, the other is to pass them through. The naming does not call out the second usage, and the method being non-const would I think be unexpected given that. 

@pranavsharma thoughts?

File: onnxruntime/core/framework/config_options.h [hunk 2]
```diff
@@ -32,7 +31,7 @@ struct ConfigOptions {
   std::string GetConfigOrDefault(const std::string& config_key, const std::string& default_value) const noexcept;
 
   // Add a config pair (config_key, config_value) to this instance of ConfigOptions
-  Status AddConfigEntry(const char* config_key, const char* config_value) noexcept;
+  common::Status AddConfigEntry(const char* config_key, const char* config_value) noexcept;
```
Comment (skottmckay) [inline onnxruntime/core/framework/config_options.h:34]: Is this change necessary?

File: include/onnxruntime/core/framework/execution_provider.h (hunk 1 above)
Comment (cloudhan) [reply include/onnxruntime/core/framework/execution_provider.h:304]: I think it would be better to let the EP copy a new one and legalize for that one only instead of the shared one. Then the behavior will not be shared between EPs.  Originally in #11138 , there is only CUDA and DML to be specially handled, so it would be a big problem at that time.

Comment (wejoncy) [reply include/onnxruntime/core/framework/execution_provider.h:304]: But what if session options of ep specific has side-effect for another EP?

Comment (cloudhan) [reply include/onnxruntime/core/framework/execution_provider.h:304]: @wejoncy Then it will be codesign scenario, the dev need to maintain both EPs.

Comment (pranavsharma) [reply include/onnxruntime/core/framework/execution_provider.h:304]: Let's take this offline in an email.

**Comment by wejoncy** [comment]
Hi @pengwa  Could you please take a look if this renaming for CudaProviderCreating has any effect on training?

File: onnxruntime/core/framework/config_options.h (hunk 2 above)
Comment (skottmckay) [reply onnxruntime/core/framework/config_options.h:34]: Unless there's a specific reason to make the code more verbose or fix a compile issue I would suggest leaving it as is. We actually do the below so that we don't need to qualify the usage of Status in the onnxruntime domain:

https://github.com/microsoft/onnxruntime/blob/565da712750ea7fd2c56ee71bae84fa557c13aab/include/onnxruntime/core/common/status.h#L192-L193

File: include/onnxruntime/core/framework/execution_provider.h [hunk 3]
```diff
@@ -299,6 +300,16 @@ class IExecutionProvider {
...
+   * Rather than put the onus on the user to know these, make the appropriate change while logging the change.
+   *
+   * @param session_options The SessionOptions object to be altered.
+   * @param logger The EP write the reason of the altering to it.
+   */
+  virtual common::Status CheckSessionOptions(const SessionOptions& /*session_options*/) {
```
Comment (skottmckay) [inline include/onnxruntime/core/framework/execution_provider.h:310]: nit: add an empty line between the previous function and the start of the comment for this one.

The comment needs updating as well as the EP should not alter the SessionOptions object.
I would suggest adding an `@return` saying the EP should return an error if it's incompatible with the SessionOptions that specifies which option is problematic so the user knows what needs to be fixed.

'Check' doesn't really convey that the EP might update its internals based on the session options, which is why this method is non-const. Maybe ProcessSessionOptions would be slightly better.

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.cpp [hunk 4]
```diff
@@ -943,6 +943,28 @@ namespace Dml
         return onnxruntime::common::Status::OK();
     }
 
+    onnxruntime::common::Status ExecutionProviderImpl::CheckSessionOptions(const onnxruntime::SessionOptions& so)
+    {
+        // DML's memory is not byte addressable and hence mem pattern doesn't work.
+        if (so.enable_mem_pattern) {
+            return ORT_MAKE_STATUS(ONNXRUNTIME,
+                                FAIL,
```
Comment (skottmckay) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.cpp:951]: nit: indentation is off. please use clang-format

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc [hunk 5]
```diff
@@ -87,14 +88,38 @@ std::unique_ptr<KernelRegistry> RegisterKernels() {
...
   }
 }
 
+common::Status XnnpackExecutionProvider::CheckSessionOptions(const SessionOptions& so) {
+  if (so.execution_mode != ExecutionMode::ORT_SEQUENTIAL) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
+                           "Parallel execution mode does not support the Xnnpack Execution Provider. "
+                           "Please set the execution mode as sequential for this "
+                           "session since it uses the Xnnpack Execution Provider.");
+  }
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:105]: What's the reason for this? For CUDA it's due to how the CUDA streams work. What about the xnnpack kernel execution would not work with the parallel executor?

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc [hunk 6]
```diff
@@ -87,14 +88,38 @@ std::unique_ptr<KernelRegistry> RegisterKernels() {
...
+  if (so.intra_op_param.allow_spinning && so.intra_op_param.thread_pool_size > 1) {
+    LOGS_DEFAULT(WARNING)
+        << "XNNPACK EP utilize pthreadpool for multi-threading. So, if allow_spinning on ORT's"
+           "thread-pool and its pool size is not 1, "
+           "pthreadpool will content with ORT's intra-op thread pool and hurt performance a lot. "
+           "Please Setting intra_op_param.allow_spinning to false or "
+           "setting ort's pool size (intra_thread_num) to 1 and try again.";
+  }
+  // 0 means user didn't set the value.
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:114]: nit: leave a space between completely separate blocks so it's easier to skim between them. when they're joined together it takes time to process whether it's a new 'if' or a continuation of the previous one (like an 'else if' or 'if').

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc [hunk 7]
```diff
@@ -87,14 +88,38 @@ std::unique_ptr<KernelRegistry> RegisterKernels() {
...
+common::Status XnnpackExecutionProvider::CheckSessionOptions(const SessionOptions& so) {
+  if (so.execution_mode != ExecutionMode::ORT_SEQUENTIAL) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
+                           "Parallel execution mode does not support the Xnnpack Execution Provider. "
+                           "Please set the execution mode as sequential for this "
+                           "session since it uses the Xnnpack Execution Provider.");
+  }
+  if (so.intra_op_param.allow_spinning && so.intra_op_param.thread_pool_size > 1) {
+    LOGS_DEFAULT(WARNING)
+        << "XNNPACK EP utilize pthreadpool for multi-threading. So, if allow_spinning on ORT's"
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:108]: suggestion:
The XNNPACK EP utilizes an internal pthread-based thread pool for multi-threading. 
If ORT's thread pool size is > 1 and spinning is enabled there will be contention between the two thread pools, and performance will suffer.
Please set either intra_op_param.allow_spinning to 0 in the SessionOption config params, or the ORT intra-op threadpool size to 1 and try again.

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc [hunk 8]
```diff
@@ -87,14 +88,38 @@ std::unique_ptr<KernelRegistry> RegisterKernels() {
...
+           "thread-pool and its pool size is not 1, "
+           "pthreadpool will content with ORT's intra-op thread pool and hurt performance a lot. "
+           "Please Setting intra_op_param.allow_spinning to false or "
+           "setting ort's pool size (intra_thread_num) to 1 and try again.";
+  }
+  // 0 means user didn't set the value.
+  if (so.intra_op_param.thread_pool_size > 1 && info_.xnn_thread_pool_size == 0) {
+    LOGS_DEFAULT(INFO) << "XNNPACK pool size is not set. Using ORT's thread-pool size as default:";
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:116]: nit: include the size in the message

File: onnxruntime/test/framework/inference_session_test.cc [hunk 9]
```diff
@@ -2068,19 +2068,9 @@ TEST(InferenceSessionTests, TestParallelExecutionWithCudaProvider) {
...
-
-  ASSERT_TRUE(status.IsOK());
-
-  const auto& so_queried = session_object.GetSessionOptions();
+  auto status = (session_object.RegisterExecutionProvider(DefaultCudaExecutionProvider()));
 
-  // execution mode is sequential since we have registered the CUDA EP
-  // (which isn't supported by the parallel execution mode)
-  ASSERT_TRUE(so_queried.execution_mode == ExecutionMode::ORT_SEQUENTIAL);
+  ASSERT_FALSE(status.IsOK());
```
Comment (skottmckay) [inline onnxruntime/test/framework/inference_session_test.cc:2073]: Is there any point to keeping this test?

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc [hunk 10]
```diff
@@ -87,14 +88,38 @@ std::unique_ptr<KernelRegistry> RegisterKernels() {
...
-    : IExecutionProvider{kXnnpackExecutionProvider, true} {
+    : IExecutionProvider{kXnnpackExecutionProvider, true}, info_(info) {
   if (info.xnn_thread_pool_size > 1) {
     // pthreadpool is independent of ort-threadpoool, so we have to disable cpu spinning for ort-threadpool.
     // otherwise, the pthreadpool will be starved and harm performance a lot.
     xnnpack_thread_pool_ = pthreadpool_create(static_cast<size_t>(info.xnn_thread_pool_size));
   }
 }
 
+common::Status XnnpackExecutionProvider::CheckSessionOptions(const SessionOptions& so) {
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:99]: Would be good to add a unit test that validates this new code.

File: onnxruntime/core/framework/config_options.h (hunk 2 above)
Comment (wejoncy) [reply onnxruntime/core/framework/config_options.h]: I didn't notice this declaration in the below of status.h.
Removed.

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc (hunk 5 above)
Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:105]: For all xnn_setup_xxxx like function, it will change xnn_operator_t.

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc (hunk 8 above)
Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:116]: Done

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc (hunk 6 above)
Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:114]: Done

File: onnxruntime/test/framework/inference_session_test.cc (hunk 9 above)
Comment (wejoncy) [reply onnxruntime/test/framework/inference_session_test.cc:2073]: you are right, this test will be covered with checksessionoption test.

File: include/onnxruntime/core/framework/execution_provider.h (hunk 3 above)
Comment (wejoncy) [reply include/onnxruntime/core/framework/execution_provider.h:310]: We do have process EP itself in this function,  so we can't mark this method as const.
This function should works with two purpose: 1 check if EP could work SessionOptions, 2. Adjust EP-self to make compatiable to run normally.  The first one is  compulsory, the second is not. How about combine them together? say `CheckSessionOptionsAndProcess`

Comment (skottmckay) [reply include/onnxruntime/core/framework/execution_provider.h]: 'Process' is fairly generic. 'checking' is a form of processing, as is updating say threadpool sizes. I don't think anything is 'compulsory' - it's up to the EP to decide what it wants to do, so I'm not sure I see value in the longer name. If it was compulsory I would have expected every single EP was required to implement it.

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc (hunk 5 above)
Comment (skottmckay) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:105]: We prevent concurrent calls to Run to avoid an individual node being executed concurrently for that reason. However, the parallel executor is not going to execute an individual node concurrently - it will potentially execute different nodes concurrently which would all be separate instances of the kernel.

File: include/onnxruntime/core/framework/execution_provider.h (hunk 3 above)
Comment (pranavsharma) [reply include/onnxruntime/core/framework/execution_provider.h]: Can we rename it to ValidateSessionOptions?

File: include/onnxruntime/core/framework/execution_provider.h [hunk 11]
```diff
@@ -300,6 +301,19 @@ class IExecutionProvider {
...
+  /** Some session option values (default or user provided) may not work with some EPs.
+   * The EP should check and assure the SessionOptions object fit its implementation to avoid potential crash.
+   * we would rather throw errors to user and remind of setting appropriate config.
+   *
+   * @param session_options The SessionOptions object to be altered.
+   * @return Status indicating whether the SessionOptions object is valid for the EP or not.
+   * The EP should return an error if it's incompatible with the SessionOptions that specifies
+   * which option is problematic so the user knows what needs to be fixed
+   */
+  virtual common::Status CheckSessionOptionsAndProcess(const SessionOptions& /*session_options*/) {
```
Comment (pranavsharma) [inline include/onnxruntime/core/framework/execution_provider.h:313]: How does someone writing a new EP know which options they should care about to implement this?

File: onnxruntime/core/providers/cann/cann_execution_provider.cc [hunk 12]
```diff
@@ -746,4 +747,14 @@ void CANNExecutionProvider::RegisterAllocator(AllocatorManager& allocator_manage
   }
 }
 
+common::Status CANNExecutionProvider::CheckSessionOptionsAndProcess(const SessionOptions& so) {
+  if (so.execution_mode != ExecutionMode::ORT_SEQUENTIAL) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
+                           "Parallel execution mode does not support the CANN Execution Provider. "
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cann/cann_execution_provider.cc:753]: This line is a bit misleading. May be we can say that parallel execution mode is incompatible with this execution provider?

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc [hunk 13]
```diff
@@ -87,14 +88,41 @@ std::unique_ptr<KernelRegistry> RegisterKernels() {
...
+           "Please set either intra_op_param.allow_spinning to 0 in the SessionOption config params,"
+           "or the ORT intra-op threadpool size to 1 and try again.";
+  }
+
+  // 0 means user didn't set the value.
+  if (so.intra_op_param.thread_pool_size > 1 && info_.xnn_thread_pool_size == 0) {
+    LOGS_DEFAULT(INFO) << "XNNPACK pool size is not set. Using ORT's thread-pool size as default:"
+                       << so.intra_op_param.thread_pool_size;
+    info_.xnn_thread_pool_size = so.intra_op_param.thread_pool_size;
+    xnnpack_thread_pool_ = pthreadpool_create(static_cast<size_t>(info_.xnn_thread_pool_size));
```
Comment (pranavsharma) [inline onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:121]: Do we really want to create a threadpool in a validation function?

Comment (skottmckay) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:121]: This is why I think ProcessSessionOptions is a better name as this is more than a 'validation' function. I don't think there's another place to create the threadpool given you don't know the ideal default size until you see the session options.

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc (hunk 5 above)
Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:105]: Got it.

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc (hunk 13 above)
Comment (pranavsharma) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:121]: I think the ideal fix would be for CreateProvider to accept a SessionOptions parameter, but I understand that this would involve a lot of changes. If there's no hurry we should do it. Otherwise, would the following work instead for the short term - store the SessionOptions ptr when creating the xnnpack EP factory and use it inside its CreateProvider method? This avoids introducing a new EP API.

File: include/onnxruntime/core/framework/execution_provider.h (hunk 11 above)
Comment (wejoncy) [reply include/onnxruntime/core/framework/execution_provider.h:313]: I think it's EP author's onus to figure out which options is not compatiable to this new EP. After all, we can't rely on the framework or ORT users to take this responsibility.
A new EP should handle all conflicts  with sessionOption, to either acommdate EP-self or report a error.

May we could document all sessionOptions in the how-to-guide, `add a new EP`.

File: onnxruntime/core/providers/cann/cann_execution_provider.cc (hunk 12 above)
Comment (wejoncy) [reply onnxruntime/core/providers/cann/cann_execution_provider.cc:753]: Sounds good.

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc (hunk 13 above)
Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:121]: I have tried to create a new API to accept a SessionOptions parameter. But it really involed too many changes especialy for ort-training.
A flaw of this new API is that we can't prevent users modify SessionOptions again after CreateProviders. It makes options in both EP and inferencesession inconsistent.

Comment (pranavsharma) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:114]: I'm not suggesting changing CreateProvider. For the short term you can change ```XnnpackProviderFactory(const ProviderOptions& provider_options)``` to accept SessionOptions. Store than inside the factory obj and use it in the implementation of CreateProvider.

Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:114]: Sounds viable. But I have to put off the operation of creat_threadpool to `GetCapability`.

But there is a special cases,  `SessionOptions` could be stored in models. Can our workaround cover that?

A  EP function doing someting according to SesionOption seems not bad,  say to do a adjustment dynamiclly.

Comment (pranavsharma) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:114]: You could create the threadpool in the xnnpack constructor. SessionOptions won't be stored in models, so you can safely ignore that.

Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:114]: In this way, It's not easy to keep consistentcy of sessionOption  between XNNPACK EP and InferenceSession.  Just like I mentioned, It's possible to alter SessionOptions after EP is created. In that case, It's even hard to debug.

we can create EP alone and then create InferenceSession. Any code could be inserted into the middle of them.

Comment (skottmckay) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:114]: I don't think a user should have an expectation that a change to the SessionOptions made after the InferenceSession is created will be applied.

Comment (pranavsharma) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:114]: EP instance creation takes place inside the CreateSession API call (not when you append an EP to SessionOptions) and SessionOptions would've been finalized by the user by then. If they change SessionOptions after that, well, that's a user error.

Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:114]: Got it.

File: include/onnxruntime/core/framework/execution_provider.h (hunk 3 above)
Comment (wejoncy) [reply include/onnxruntime/core/framework/execution_provider.h]: Done.

Comment (pranavsharma) [reply include/onnxruntime/core/framework/execution_provider.h]: I guess we don't need this API any more since we agreed to go with the factory based solution, right?

Comment (wejoncy) [reply include/onnxruntime/core/framework/execution_provider.h]: Not quite sure if this condition based on EP-in-factory  solution works as expected.

For example, we can create factory and EP with python-binding
https://github.com/microsoft/onnxruntime/blob/aefc82b5c99671b527ff0196cce9c5d002d71343/onnxruntime/python/onnxruntime_pybind_schema.cc#L65
However, it didn't touch any sessionOptions here, which means we have to pass nullptr to `xxFactory::create`, Then we create EP successfully, But it will lead to troubles or even crash.

Any thoughs on this question? Thanks

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 14]
```diff
@@ -3500,6 +3500,7 @@ struct OrtApi {
    *
    * XNNPACK supported keys:
    *   "intra_op_num_threads": number of thread-pool size to use for XNNPACK execution provider.
+   *      default value is 0, which means to use the global thread-pool size.
```
Comment (pranavsharma) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:3503]: Global has a different meaning. Did you mean "session" here?

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc [hunk 15]
```diff
@@ -123,10 +124,27 @@ using namespace xnnpack;
...
 XnnpackExecutionProvider::XnnpackExecutionProvider(const XnnpackExecutionProviderInfo& info)
     : IExecutionProvider{kXnnpackExecutionProvider, true} {
-  if (info.xnn_thread_pool_size > 1) {
-    // pthreadpool is independent of ort-threadpoool, so we have to disable cpu spinning for ort-threadpool.
-    // otherwise, the pthreadpool will be starved and harm performance a lot.
-    xnnpack_thread_pool_ = pthreadpool_create(static_cast<size_t>(info.xnn_thread_pool_size));
+  int xnn_thread_pool_size = info.xnn_thread_pool_size;
+  int ort_thread_pool_size = info.session_options ? info.session_options->intra_op_param.thread_pool_size : 1;
+  bool allow_intra_op_spinning =
+      info.session_options && info.session_options->config_options.GetConfigOrDefault(
```
Comment (pranavsharma) [inline onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc:130]: Is it expected to turn off spinning if session options is null? In the regular inferencing workflow spinning is turned on by default.

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 14 above)
Comment (wejoncy) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:3503]: I see. ORT can have multiple session-configs

Comment (pranavsharma) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:3503]: It's because we've the concept of a global threadpool and a session threadpool. Here you mean session since you're passing session options.

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc (hunk 15 above)
Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc]: Thanks. It's reasonable to keep alignment with the default value.

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 14 above)
Comment (wejoncy) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: Got it. No wonder the web side writes it only takes global threadpool.

**Review by pranavsharma — APPROVED** [review]

**Comment by pranavsharma** [comment]
Please update PR description to reflect the latest changes as it seems out of date.

**Comment by wejoncy** [comment]
> Please update PR description to reflect the latest changes as it seems out of date.

sure
